tutorial>alerting:get started with grafana alerting part 3 (#98093)
grafana/grafana: docs/sources/tutorials/alerting-get-started-pt3/index.md

  alert-grouping:
    - pattern: /docs/grafana/
      destination: /docs/grafana/<GRAFANA_VERSION>/alerting/fundamentals/notifications/group-alert-notifications/
    - pattern: /docs/grafana-cloud/
      destination: /docs/grafana-cloud/alerting-and-irm/alerting/fundamentals/notifications/group-alert-notifications/
---

<!-- INTERACTIVE page intro.md START -->

# Get started with Grafana Alerting - Part 3

The Get started with Grafana Alerting tutorial Part 3 is a continuation of [Get started with Grafana Alerting tutorial Part 2](http://www.grafana.com/tutorials/alerting-get-started-pt2/).

Alert grouping in Grafana Alerting reduces notification noise by combining related alerts into a single, concise notification. This is essential for on-call engineers, ensuring they focus on resolving incidents instead of sorting through a flood of notifications.

Grouping is configured by using labels in the notification policy that reference the labels that are generated by the alert instances. With notification policies, you can also configure how often notifications are sent for each group of alerts.

In this tutorial, you will:

> - Understand how alert rule grouping works.

Reviewer: ```suggestion
- Learn how alert rule grouping works.
```

   {{< docs/ignore >}}

   NOTE:

   If you already have Grafana, Loki, or Prometheus running on your system, you might see errors, because the Docker image is trying to use ports that your local installations are already using. If this is the case, stop the services, then run the command again.

   {{< /docs/ignore >}}

<!-- INTERACTIVE page step1.md END -->
<!-- INTERACTIVE page step2.md START -->

## How alert rule grouping works

Alert notification grouping is configured with **labels** and **timing options**:

- **Labels** map the alert rule with the notification policy and define the grouping.
- **Timing options** control when and how often notifications are sent.

{{< figure src="/media/docs/alerting/alerting-notification-policy-diagram-with-labels-v3.png" max-width="750px" alt="A diagram about the components of a notification policy, including labels and groups" >}}

### Types of Labels

1. **Reserved labels** (default):

   - Automatically generated by Grafana, e.g., `alertname`, `grafana_folder`.
   - Example: `alertname="High CPU usage"`.

1. **User-configured labels**:

   - Added manually to the alert rule.
   - Example: `severity`, `priority`.

> 1. **Query labels**:

Reviewer: add a space here to be consistent with the above?

   - Returned by the data source query.
   - Example: `region`, `service`, `environment`.

### Timing Options

1. **Group wait**: Time before sending the first notification.
1. **Group interval**: Time between notifications for a group.
1. **Repeat interval**: Time before resending notifications for an unchanged group.

> Alerts sharing the **same label values** are grouped together, and timing options determine notification frequency.

Reviewer: alerts .... instances / rules ...?

For more details, see:

- [Grouping Alerts](ref:alert-grouping)
- [Alert Labels](ref:alert-labels)

<!-- INTERACTIVE page step2.md END -->
<!-- INTERACTIVE page step3.md START -->

> ## A real-world example of alert grouping in action

Reviewer: you need to add some text under this header - like an intro line. we don't have sequential headers

### Scenario: monitoring a distributed application

You’re monitoring metrics like CPU usage, memory utilization, and network latency across multiple regions. Alert rules include labels such as `region: us-west` and `region: us-east`. If multiple alerts trigger across these regions, they can result in notification floods.

### How to manage grouping

To group alert rule notifications:

1. **Define labels**: Use `region`, `metric`, or `instance` labels to categorize alerts.
1. **Configure Notification policies**:
   - Group alerts by the `region` label.
   - Example:
     - Alerts for `region: us-west` go to the West Coast team.
     - Alerts for `region: us-east` go to the East Coast team.

<!-- INTERACTIVE page step3.md END -->
<!-- INTERACTIVE page step4.md START -->

> ## Setting up alert rule grouping

Reviewer: ```suggestion
## Set up alert rule grouping
```

Reviewer: same here - some intro text before the next header

1. Override grouping settings:

   - Toggle **Override grouping**.
   - **Group by**: `region`.

     **Group by** consolidates alerts that share the same grouping label into a single notification. For example, all alerts with `region=us-west` will be combined into one notification, making it easier to manage and reducing alert fatigue.

1. Set custom timing:

   - Toggle **Override general timings**.
   - **Group interval**: `2m`. This ensures follow-up notifications for the same alert group will be sent at intervals of 2 minutes.

     **Timing options** control how often notifications are sent and can help balance timely alerting with minimizing noise.

1. Save and repeat:

   - Repeat for `region = us-east` with a different webhook or a different contact point.

   {{< figure src="/media/docs/alerting/notificaiton-policies-region.png" max-width="750px" alt="Two nested notification policies to route and group alert notifications" >}}

   These nested policies should route alert instances where the region label is either us-west or us-east.

>    **Note**: In Grafana, each label within a notification policy must have a unique key. If you attempt to add the same label key (e.g., region) with different values (us-west and us-east), only the last entry will be saved, and the previous one will be discarded. This is because labels are stored as associative arrays (maps), where each key must be unique.

Reviewer: Use the format for notes?

   For identical label keys use regex matchers (e.g., _region=~"us-west|us-east"_).

<!-- INTERACTIVE ignore END -->

{{< docs/ignore >}}

1. Visit [http://localhost:3000](http://localhost:3000), where Grafana should be running
1. Navigate to **Alerts & IRM > Alerting > Notification policies**.
1. In the Default policy, click **+ New child policy**.

   - In the Default policy, click **+ New child policy**.
   - **Label**: `region`
   - **Operator**: `=`
   - **Value**: `us-west`

>    This label matches alert rules where the region label is us-west

Reviewer: ```suggestion
   This label matches alert rules where the region label is us-west.
```

1. Choose a **Contact point**:

   - Select **Webhook**.

   If you don’t have any contact points, add a Contact point.

1. Enable Continue matching:

   - Turn on **Continue matching subsequent sibling nodes** so the evaluation continues even after one or more labels (i.e. region label) match.

1. Override grouping settings:

   - Toggle **Override grouping**.
   - **Group by**: `region`.

     **Group by** consolidates alerts that share the same grouping label into a single notification. For example, all alerts with `region=us-west` will be combined into one notification, making it easier to manage and reducing alert fatigue.

1. Set custom timing:

   - Toggle **Override general timings**.
   - **Group interval**: `2m`. This ensures follow-up notifications for the same alert group will be sent at intervals of 2 minutes.

     **Timing options** control how often notifications are sent and can help balance timely alerting with minimizing noise.

1. Save and repeat:

   - Repeat for `region = us-east` with a different webhook or a different contact point.

>      **Note**: In Grafana, each label within a notification policy must have a unique key. If you attempt to add the same label key (e.g., region) with different values (us-west and us-east), only the last entry will be saved, and the previous one will be discarded. This is because labels are stored as associative arrays (maps), where each key must be unique.

Reviewer: note format

>      **Note**: In Grafana, each label within a notification policy must have a unique key. If you attempt to add the same label key (e.g., region) with different values (us-west and us-east), only the last entry will be saved, and the previous one will be discarded. This is because labels are stored as associative arrays (maps), where each key must be unique.

Reviewer: ```suggestion
     **Note**: In Grafana, each label within a notification policy must have a unique key. If you attempt to add the same label key (e.g., region) with different values (us-west and us-east), only the last entry is saved, and the previous one is discarded. This is because labels are stored as associative arrays (maps), where each key must be unique.
```

     For identical label keys use regex matchers (e.g., _region=~"us-west|us-east"_).

{{< /docs/ignore >}}

<!-- INTERACTIVE page step4.md END -->
<!-- INTERACTIVE page step5.md START -->

## Create an alert rule

In this section we configure an alert rule based on our application monitoring example.

> 1. Go to **Alerting > Alert rules**.

Reviewer: ```suggestion
1. Navigate to **Alerting > Alert rules**.
```

2. Click **New alert rule**.

### Enter an alert rule name

> Make it short and descriptive as this will appear in your alert notification. For instance, `High CPU usage - Multi-region`.

Reviewer: ```suggestion
Make it short and descriptive as this appears in your alert notification. For instance, `High CPU usage - Multi-region`.
```

### Define query and alert condition

In this section, we use the default options for Grafana-managed alert rule creation. The default options let us define the query, a expression (used to manipulate the data -- the `WHEN` field in the UI), and the condition that must be met for the alert to be triggered (in default mode is the threshold).

Grafana includes a [test data source](https://grafana.com/docs/grafana/latest/datasources/testdata/) that creates simulated time series data. This data source is included in the demo environment for this tutorial. If you're working in Grafana Cloud or your own local Grafana instance, you can add the data source through the **Connections** menu.

> 1. Select **TestData** data source from the drop-down menu.

Reviewer: ```suggestion
1. From the drop-down menu, select **TestData** data source.
```

   These nested policies should route alert instances where the region label is either us-west or us-east.

   {{< admonition type="note" >}}

>    **Note**: In Grafana, each label within a notification policy must have a unique key. If you attempt to add the same label key (e.g., region) with different values (us-west and us-east), only the last entry is saved, and the previous one is discarded. This is because labels are stored as associative arrays (maps), where each key must be unique.

Reviewer: ```suggestion
   In Grafana, each label within a notification policy must have a unique key. If you attempt to add the same label key (e.g., region) with different values (us-west and us-east), only the last entry is saved, and the previous one is discarded. This is because labels are stored as associative arrays (maps), where each key must be unique.
```

Reviewer: always take out the **Note** because the admonition adds it anyway

Author: good catch . thank you!
